Assignment 2.Rmd

```diff
@@ -94,7 +144,64 @@ In Part II your task is to [look up](http://igraph.org/r/) in the igraph documen
 Now practice with data from our class. Please create a **person-network** with the data set hudk4050-classes.csv. To create this network you will need to create a person-class matrix using the tidyr functions and then create a person-person matrix using `t()`. You will then need to plot a matrix rather than a data frame using igraph.
 
 Once you have done this, also [look up](http://igraph.org/r/) how to generate the following network metrics: betweeness centrality and dregree. **Who is the most central person in the network?**
+```{r}
+library(tidyr)
+library(visNetwork)
+# read original csv and name as DF1
+DF1 <- read.csv("hudk4050-classes.csv", header = TRUE, na.strings = "")
+# choose what we want from DF1 and name as DF2: take columns of class 1 ~ class 6 and collapses into key-value pairs, key is "course", value is "code" of course; paste together columns of first name and last name into one column called "name"; keep the column of name and code
+DF2 <- DF1 %>% 
+  gather(course, code, Class.1: Class.6) %>%
+  unite(name, First.Name, Last.Name, sep = "_") %>%
+  select(name, code) 
+# remove spaces between letters and numbers in the "code" column
+DF2$code <- gsub("", "", DF2$code)
+# prepare DF3 ready for matrix: count 1 if the person takes this course; remove all HUDK4050 in case the matrix will be too complicated to read; spread key-value pairs, key is the code, value is the count 
+DF3 <- DF2 %>% 
+  mutate(count = 1) %>%
+  filter(code != "HUDK4050") %>%
+  spread(code, count)
+# create a matrix of DF3
+DF3_matrix <- data.matrix(DF3)
+# set rownames as person name
+rownames(DF3_matrix) <- DF3$name
+# fill the matrix with 0 & 1 according to whether the cell is NA
+DF3_matrix <- ifelse(is.na(DF3_matrix), 0, 1)
+# create a final matrix which connect students and courses
+finalMatrix <- DF3_matrix %*% t(DF3_matrix)
+# remove all diagonols in case all the students connect to themselves
+diag(finalMatrix) <- 0
+# plot graph
+graph <- graph.adjacency(finalMatrix, mode = "undirected")
+plot(graph, vertex.size=5, vertex.color="green")
+#create a more pretty version of the graph using visNetwork
+data <- toVisNetworkData(graph)
+nodes <- data.frame(data$nodes,
+  # add labels on nodes
+  label = DF3$name,
+  color = "orange")   
+visNetwork(nodes, edges = data$edges, height = "500px", width = "100%")
+```
+
+```{r}
+# create matrix of degree and betweenness
+centralMatrix <- DF3 %>% 
+  mutate(degree = degree(graph, v=V(graph))) %>%
+  mutate(betweenness = betweenness(graph, v=V(graph))) %>%
+  select(name, degree, betweenness)
+# create a matrix called centralMatrix which is sorted by degree
+centralMatrix <- centralMatrix %>% arrange(desc(degree))
+# print centralMatrix
+centralMatrix
+```
+# The first student (Lintong Li) in centralMetrix is the central person in our class.
+
```

Comment:
Do you understand the difference between betweeness and degree centrality? They indicate different aspects of centrality. Degree = how many connections a node has, betweeness = the number of shortest paths that pass through a node.